src/librustc_mir/transform/unreachable_prop.rs

```diff
@@ -0,0 +1,37 @@
+//! A pass that propagates the unreachable terminator of a block to its predecessors
+//! when all of their successors are unreachable. This is achieved through a
+//! post-order traversal of the blocks.
+
+use rustc::ty::TyCtxt;
+use rustc::mir::*;
+use rustc_data_structures::fx::FxHashSet;
+use crate::transform::{MirPass, MirSource};
+
+pub struct UnreachablePropagation;
+
+impl MirPass<'_> for UnreachablePropagation {
+    fn run_pass<'tcx>(&self, _tcx: TyCtxt<'tcx>, _: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        let mut unreachable_blocks = FxHashSet::default();
+
+        for (bb, bb_data) in traversal::postorder(body) {
+            if let Some(terminator) = &bb_data.terminator {
+                if let TerminatorKind::Unreachable = terminator.kind {
+                    unreachable_blocks.insert(bb.clone());
+                } else {
+                    let mut bb_successors = terminator.successors().peekable();
+
+                    if bb_successors.peek().is_some() {
+                        if bb_successors.all(|bb| unreachable_blocks.contains(bb)) {
```

Comment:
We can generalize this by only retaining the reachable blocks.
For example, if you consider `SwitchInt(_) -> [bb2, bb3, otherwise: bb4]` and `bb3` is unreachable,
then you could transform that to `SwitchInt(_) -> [bb2, otherwise: bb4]`.

Reply:
Here's an idea for doing this in the most general way: Define a `retain_successors` function taking a `predicate` which handles the various forms of `TerminatorKind`s removing a successor if the predicate does not hold. If there is just one successor, the terminator itself becomes `Goto`, if none, the terminator becomes `Unreachable`.

Reply:
Ok! Will do!